lib/install/management/finder.go

```diff
@@ -281,22 +280,59 @@ func (d *Dispatcher) searchVCHsPerDC(dc *object.Datacenter) ([]*vm.VirtualMachin
 	d.session.Finder.SetDatacenter(dc)
 
 	var vchs []*vm.VirtualMachine
-	if pools, err = d.session.Finder.ResourcePoolList(d.op, "*"); err != nil {
+
+	pools, err = d.listResourcePools("*")
+	if err != nil {
+		err = errors.Errorf("Failed to search resource pools for datacenter %q: %s", dc.InventoryPath, err)
+		return nil, err
+	}
+
+	vchs = d.searchVCHsFromPools(pools)
+	return vchs, nil
+}
+
+// listResourcePool outputs a list of all resource pools under a compute path
+// retries on ObjectNotFound error because sometimes it's due to concurrent modifications of resource pools
+func (d *Dispatcher) listResourcePools(path string) ([]*object.ResourcePool, error) {
+	var err error
+	var pools []*object.ResourcePool
+
+	pools, err = d.session.Finder.ResourcePoolList(d.op, path)
+
+	// under some circumstances, such as when there is concurrent vic-machine elete operation running in the background,
+	// listing resource pools might fail because some VCH pool is being destroyed at the same time.
+	// If that happens, we retry and list pools again
+	err = retry.Do(func() error {
+		pools, err = d.session.Finder.ResourcePoolList(d.op, path)
 		if _, ok := err.(*find.NotFoundError); ok {
-			return vchs, nil
+			return nil
 		}
-		err = errors.Errorf("Failed to search resource pools for datacenter %q: %s", dc.InventoryPath, err)
+		return err
+	}, func(err error) bool {
+		return tasks.WrapRetryError(d.op, err, tasks.IsNotFoundError)
+	})
+
+	if err != nil {
 		return nil, err
 	}
 
+	return pools, nil
```

Comment:
I think this:
```go
if err != nil {
    return nil, err
}
return pools, nil
```

Could be simplified to:
```go
return pools, err
```

(Either `pools` or `err` should be `nil`.)

Reply:
good catch, thanks

Reply:
side story: also I was kind of struggling on the possibility that both `pools` and `err` are non-nil...I was wondering if this could happen when there was error listing one/some of the pools but we still want to get the rest of the pools. But it turned out that we don't get any pool back when searching one of the pools failed...
it feels like not the expected behavior because any listing operation shouldn't abort if one of the items failed to search tho...